core/src/main/scala/org/apache/spark/util/ListenerBus.scala

```diff
@@ -17,53 +17,222 @@
+            val prevLastReportTimestamp = lastReportTimestamp
+            lastReportTimestamp = System.currentTimeMillis()
+            logError(s"Dropping $droppedEvents SparkListenerEvent since " +
+              new java.util.Date(prevLastReportTimestamp) +
+              " because no remaining room in event queue. This likely means" +
+              s" $listenerName event processor is too slow and cannot keep up " +
+              "with the rate at which tasks are being started by the scheduler.")
+          }
+        }
+    }
+  }
+
+  private[this] def isProcessingEvent: Boolean = processingEvent.get()
+
+  def isEmpty: Boolean = {
+    executorService.getQueue.size() == 0 && !isProcessingEvent
+  }
+
+  def stop(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10, TimeUnit.SECONDS)
+  }
+}
 
 /**
  * An event bus which posts events to its listeners.
  */
 private[spark] trait ListenerBus[L <: AnyRef, E] extends Logging {
 
-  // Marked `private[spark]` for access in tests.
-  private[spark] val listeners = new CopyOnWriteArrayList[L]
+  // Cap the capacity of the event queue so we get an explicit error (rather than
+  // an OOM exception) if it's perpetually being added to more quickly than it's being drained.
+  protected def eventQueueSize = 10000
+  private val eventGroupToEventExecutors =
+    new ConcurrentHashMap[String, ListenerEventExecutor[L]] ()
+
+  // Indicate if `start()` is called
+  private val started = new AtomicBoolean(false)
+  // Indicate if `stop()` is called
+  private val stopped = new AtomicBoolean(false)
 
   /**
    * Add a listener to listen events. This method is thread-safe and can be called in any thread.
    */
-  final def addListener(listener: L): Unit = {
-    listeners.add(listener)
+  final def addListener(
+    listener: L, eventListenerGroup: String = ListenerEventExecutor.DefaultEventListenerGroup):
+  Unit = synchronized {
+    var listenerEventExecutor = eventGroupToEventExecutors.get(eventListenerGroup)
+    if (listenerEventExecutor == null) {
+      listenerEventExecutor =
+        new ListenerEventExecutor[L](listener.getClass.getName, eventQueueSize)
+      eventGroupToEventExecutors.put(eventListenerGroup, listenerEventExecutor)
+
+    }
+    listenerEventExecutor.addListener(listener)
+    if (started.get()) {
+      listenerEventExecutor.start
+    }
   }
 
   /**
    * Remove a listener and it won't receive any events. This method is thread-safe and can be called
    * in any thread.
    */
   final def removeListener(listener: L): Unit = {
-    listeners.remove(listener)
+    for (eventExecutor <- eventGroupToEventExecutors.values().asScala) {
+      eventExecutor.removeListener(listener)
+    }
+  }
+
+   /**
+    * For testing only. Returns whether there is any event pending to be processed by
+    * any of the existing listener
+    */
+  def isListenerBusEmpty: Boolean = {
+    for (eventExecutor <- eventGroupToEventExecutors.values().asScala) {
```

Comment:
I think you can do this with `eventGroupToEventExecutors.values().asScala.forall`.